ugs-core/src/com/willwinder/universalgcodesender/visualizer/GcodeViewParse.java

```diff
@@ -170,7 +172,18 @@ private static GcodeParser getParser(double arcSegmentLength) {
                 for (GcodeMeta meta : points) {
                     if (meta.point != null) {
                         addLinesFromPointSegment(start, meta.point, arcSegmentLength, lines);
-                        start.set(meta.point.point());
+                        // if the last set point is in a different or unknown unit, crate a new point-instance with the correct unit set
+                        if (start.getUnits() != UnitUtils.Units.MM && gp.getCurrentState().isMetric){
+                            start=new Position(
+                                    meta.point.point().x,
+                                    meta.point.point().y,
+                                    meta.point.point().z,
+                                    gp.getCurrentState().isMetric ? UnitUtils.Units.MM : UnitUtils.Units.INCH
+                                    );
+                        } else {
+                            // ...otherwise recycle the old instance and just update the x,y,z coords
+                            start.set(meta.point.point());
+                        }
```

Comment:
I thought the visualizer converts everything into MM, did you have a test case that this fixes?

Reply:
Only via a manual debug test, i noticed that the start-point has no unit set:

![grafik](https://user-images.githubusercontent.com/1370302/56838895-37a25980-6880-11e9-8d85-c6ed6879f79f.png)

Reply:
And with the fix applied: 
![grafik](https://user-images.githubusercontent.com/1370302/56839051-037b6880-6881-11e9-841d-beeb507deafe.png)

But I think currently this does not really change anything, I just thought it would be more correct to have the correct units here